pm/recipient_test.go

```diff
@@ -0,0 +1,851 @@
+	_, err = r.ReceiveTicket(ticket1, sig, params.Seed)
+	if err == nil {
+		t.Error("expected invalid senderNonce (new nonce = highest seen nonce) error")
+	}
+	if err != nil && !strings.Contains(err.Error(), "invalid ticket senderNonce") {
+		t.Errorf("expected invalid senderNonce (new nonce = highest seen nonce) error, got %v", err)
+	}
+
+	// Replay senderNonce = 0 (new nonce < highest seen nonce)
+	_, err = r.ReceiveTicket(ticket0, sig, params.Seed)
+	if err == nil {
+		t.Error("expected invalid senderNonce (new nonce < highest seen nonce) error")
+	}
+	if err != nil && !strings.Contains(err.Error(), "invalid ticket senderNonce") {
+		t.Errorf("expected invalid senderNonce (new nonce < highest seen nonce) error, got %v", err)
+	}
+}
+
+func TestReceiveTicket_ValidNonWinningTicket_Concurrent(t *testing.T) {
+	sender, b, v, ts, faceValue, winProb, sig := newRecipientFixture(t)
+
+	r, err := NewRecipient(b, v, ts, faceValue, winProb)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Config stub validator with valid non-winning tickets
+	v.SetIsValidTicket(true)
+
+	params, err := r.TicketParams(sender)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var wg sync.WaitGroup
+	var errCount uint64
+
+	for i := 0; i < 50; i++ {
+		wg.Add(1)
+
+		go func(senderNonce uint64) {
+			defer wg.Done()
+
+			ticket := newTicket(sender, params, senderNonce)
+
+			_, err := r.ReceiveTicket(ticket, sig, params.Seed)
+			if err != nil {
+				atomic.AddUint64(&errCount, 1)
+			}
+		}(uint64(i))
+	}
+
+	wg.Wait()
+
+	if errCount == 0 {
+		t.Error("expected more than zero senderNonce errors for concurrent ticket receipt")
+	}
+}
+
+func TestRedeemWinningTickets_InvalidSessionID(t *testing.T) {
+	_, b, v, ts, faceValue, winProb, _ := newRecipientFixture(t)
+
+	r, err := NewRecipient(b, v, ts, faceValue, winProb)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Config stub ticket store to fail load
+	ts.loadShouldFail = true
+
+	if err := r.RedeemWinningTickets("foo"); err == nil {
+		t.Error("expected ticket store error")
+	}
+}
+
+func TestRedeemWinningTickets_SingleTicket_GetDepositError(t *testing.T) {
+	sender, b, v, ts, faceValue, winProb, sig := newRecipientFixture(t)
+
+	r, err := NewRecipient(b, v, ts, faceValue, winProb)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	params, err := r.TicketParams(sender)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Config stub validator with valid winning tickets
+	v.SetIsValidTicket(true)
+	v.SetIsWinningTicket(true)
+	// Config stub broker with zero penalty escrow
+	b.SetPenaltyEscrow(sender, big.NewInt(0))
+
+	// Test get deposit error
+	ticket := newTicket(sender, params, 0)
+
+	won, err := r.ReceiveTicket(ticket, sig, params.Seed)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !won {
+		t.Fatal("expected valid winning ticket")
+	}
+
+	err = r.RedeemWinningTickets(ticket.RecipientRandHash.Hex())
+	if err == nil {
+		t.Error("expected get deposit error")
+	}
+}
+
+func TestRedeemWinningTickets_SingleTicket_GetPenaltyEscrowError(t *testing.T) {
```

Comment:
I'm slightly confused here because in the test `TestRedeemWinningTickets_SingleTicket_GetDepositError` I expected the deposit to be set to zero when in fact PE is set to zero, and in this test `TestRedeemWinningTickets_SingleTicket_GetPenaltyEscrowError` I expected PE to be set to zero, when in fact deposit is set to zero.
In both tests I also expected the other value to be explicitly set to some non-zero value.

wdyt?

Reply:
These tests are to test if `TicketBroker.GetDeposit()` and `TicketBroker.GetPenaltyEscrow()` return an error (i.e. if the RPC call to fetch the deposit/penalty escrow failed). In order to stub this error, `stubBroker` is configured such that if `SetDeposit()`/`SetPenaltyEscrow()` is never called for a given address, then it will return an error. Thus, for the `TestRedeemWinningTickets_SingleTicket_GetDepositError`, the deposit is never set, but the PE is set to zero so an error would be returned by `stubError` since the deposit is never set.

I can make a few updates to make this clearer:

- Use an explicit boolean flag to trigger errors for `TicketBroker.GetDeposit()`/`TicketBroker.GetPenaltyEscrow`
- Create `stubBroker` with default non-zero values for deposit and PE and explicitly set them to zero in tests if needed

Reply:
Awesome, I think both your suggestions would make these tests easier to understand!